client/client.go

```diff
@@ -1264,61 +1211,84 @@ func (c *Client) addAlloc(alloc *structs.Allocation) error {
-	c.consulService = cs
-	return err
-}
-
-// syncConsul removes services of tasks which are no longer in running state
-func (c *Client) syncConsul() {
-	sync := time.NewTicker(consulSyncInterval)
-	for {
-		select {
-		case <-sync.C:
-			// Give up pruning services if we can't fingerprint Consul
-
-			c.configLock.RLock()
-			_, ok := c.configCopy.Node.Attributes["consul.server"]
+// setupConsulSyncer creates Client-mode consul.Syncer callbacks that are
+// executed periodically.
+func (c *Client) setupConsulSyncer() error {
+	// The bootstrapFn callback handler is used to periodically poll
+	// Consul to look up the Nomad Servers in Consul.  In the event the
+	// heartbeat deadline has been exceeded and this Client is orphaned
+	// from its servers, periodically poll Consul to reattach this Client
+	// to its cluster and automatically recover from a detached state.
+	bootstrapFn := func() {
+		now := time.Now()
+		c.configLock.RLock()
+		if now.Before(c.backupServerDeadline) {
 			c.configLock.RUnlock()
-			if !ok {
-				continue
-			}
-			services := make(map[string]struct{})
-			// Get the existing allocs
-			c.allocLock.RLock()
-			allocs := make([]*AllocRunner, 0, len(c.allocs))
-			for _, ar := range c.allocs {
-				allocs = append(allocs, ar)
+			return
+		}
+		c.configLock.RUnlock()
+
+		nomadServerServiceName := c.config.ConsulConfig.ServerServiceName
+		services, _, err := c.consulSyncer.ConsulClient().Catalog().Service(nomadServerServiceName, consul.ServiceTagRpc, &consulapi.QueryOptions{AllowStale: true})
+		if err != nil {
+			c.logger.Printf("[WARN] client: unable to query service %q: %v", nomadServerServiceName, err)
+			return
+		}
+		serverAddrs := make([]string, 0, len(services))
+		for _, s := range services {
+			port := strconv.FormatInt(int64(s.ServicePort), 10)
+			addr := s.ServiceAddress
+			if addr == "" {
+				addr = s.Address
 			}
-			c.allocLock.RUnlock()
-			for _, ar := range allocs {
-				ar.taskStatusLock.RLock()
-				taskStates := copyTaskStates(ar.taskStates)
-				ar.taskStatusLock.RUnlock()
-				for taskName, taskState := range taskStates {
-					if taskState.State == structs.TaskStateRunning {
-						if tr, ok := ar.tasks[taskName]; ok {
-							for _, service := range tr.task.Services {
-								svcIdentifier := fmt.Sprintf("%s-%s", ar.alloc.ID, tr.task.Name)
-								services[service.ID(svcIdentifier)] = struct{}{}
-							}
+			serverAddrs = append(serverAddrs, net.JoinHostPort(addr, port))
+		}
+		c.rpcProxy.SetBackupServers(serverAddrs)
+	}
+	c.consulSyncer.AddPeriodicHandler("Nomad Client Fallback Server Handler", bootstrapFn)
+
+	consulServicesSyncFn := func() {
+		// Give up pruning services if we can't fingerprint our
+		// Consul Agent.
+		c.configLock.RLock()
+		_, ok := c.configCopy.Node.Attributes["consul.version"]
+		c.configLock.RUnlock()
+		if !ok {
+			return
+		}
+
+		services := make(map[string]struct{})
+		// Get the existing allocs
+		c.allocLock.RLock()
```

Comment:
getAllocRunners returns a copy. Can you check that all accessors of `c.allocs` are using the `c.allocLock`